scalanet/src/io/iohk/scalanet/peergroup/kademlia/KBuckets.scala

```diff
@@ -1,45 +1,40 @@
 package io.iohk.scalanet.peergroup.kademlia
 
-import java.util.concurrent.ConcurrentSkipListSet
+import java.time.Clock
 
 import scodec.bits.BitVector
 
-import scala.collection.JavaConverters._
-
 /**
   * Skeletal kbucket implementation.
   * @param baseId the nodes own id.
   */
-class KBuckets(val baseId: BitVector) {
-
-  private val nodeIds =
-    new ConcurrentSkipListSet[BitVector](new XorOrdering(baseId)).asScala
+class KBuckets(val baseId: BitVector, val clock: Clock) {
 
-  add(baseId)
+  private val buckets = (0 until baseId.length.toInt).map(_ => TimeSet[BitVector](clock)).toList
 
   /**
     * Find the n nodes closest to nodeId in kBuckets.
     * Return the resulting node records sorted by distance from the nodeId.
     * The indices into the kBuckets are defined by their distance from referenceNodeId.
     */
   def closestNodes(nodeId: BitVector, n: Int): List[BitVector] = {
-    // replace with routing tree described in kademlia paper...
-    nodeIds.toList.sorted(new XorOrdering(nodeId)).take(n)
+    iterator.toList.sorted(new XorOrdering(nodeId)).take(n)
```

Comment:
Also there is a one place in code where it is run as `kBuckets.closestNodes(key, 1).find(_ == key)`, could you at least optimize it there?

Reply:
I've made the implementation more efficient by sorting a single kbucket and then sorting additional buckets if extra results are needed.

Reply:
This also addresses your later comment about  `nodeId <- kBuckets.closestNodes(key, 1).find(_ == key)` since the value of `n=1` means the `closestNodes` search there will usually access/sort only one bucket.